src/com/amazon/ion/impl/lite/IonValueLite.java

```diff
@@ -969,64 +1057,121 @@ private static void writeFieldNameAndAnnotations(IonWriter writer, IonValueLite
+                iteratorStack[iteratorStackIndex] = currentIterator;
+                // BLANK: insert logic for handling the start of a container value.
+            }
+            do {
+                if (currentIterator == null) {
+                    // BLANK: insert logic for handling the end of iteration.
+                    return;
+                }
+                value = currentIterator.nextOrNull();
+                if (value == null) {
+                    // BLANK: insert logic for handling the end of a container value.
+                    iteratorStack[iteratorStackIndex] = null; // Allow this to be garbage collected
+                    currentIterator = (iteratorStackIndex == 0) ? null : iteratorStack[--iteratorStackIndex];
+                }
+            } while (value == null);
+        } while (true);
+     */
+
     private void writeToIterative(IonWriter writer, SymbolTableProvider symbolTableProvider) throws IOException {
-        Deque<Iterator<IonValue>> iteratorStack = null;
-        Iterator<IonValue> currentIterator = null;
+        IonContainerLite.SequenceContentIterator[] iteratorStack = new IonContainerLite.SequenceContentIterator[CONTAINER_STACK_INITIAL_CAPACITY];
+        int iteratorStackIndex = -1;
+        IonContainerLite.SequenceContentIterator currentIterator = null;
         IonValueLite value = this;
         do {
             writeFieldNameAndAnnotations(writer, value, symbolTableProvider);
-            IonType type = value.getType();
-            if (value.isNullValue()) {
-                writer.writeNull(type);
-            } else if (IonType.isContainer(type)) {
-                if (currentIterator != null) {
-                    if (iteratorStack == null) {
-                        iteratorStack = new ArrayDeque<>(CONTAINER_STACK_INITIAL_CAPACITY);
-                    }
-                    iteratorStack.add(currentIterator);
-                }
-                currentIterator = ((IonContainer) value).iterator();
-                writer.stepIn(type);
-            } else {
+            if ((value._flags & IS_NULL_VALUE) != 0) {
+                writer.writeNull(value.getType());
+            } else if (!(value instanceof IonContainerLite)) {
                 value.writeBodyTo(writer, symbolTableProvider);
+            } else {
+                if (++iteratorStackIndex >= iteratorStack.length) {
+                    iteratorStack = growIteratorStack(iteratorStack);
+                }
+                currentIterator = ((IonContainerLite) value).new SequenceContentIterator(0, true);
+                iteratorStack[iteratorStackIndex] = currentIterator;
+                writer.stepIn(value.getType());
             }
             do {
                 if (currentIterator == null) {
                     return;
                 }
-                value = currentIterator.hasNext() ? (IonValueLite) currentIterator.next() : null;
+                value = currentIterator.nextOrNull();
```

Comment:
`hashNext()` and `next()` contain a lot of redundant logic that checks for mutations. Since no mutations can possibly occur during a single call to writeTo or hashCode, we can use a single method that eliminates all of this, resulting in speedups of up to 21%.